src/multithreading/ch04/scheduler/dispatcher.hpp

```diff
@@ -0,0 +1,36 @@
+#ifndef DISPATCHER_HPP
+#define DISPATCHER_HPP
+
+#include "abstract_request.hpp"
+#include "worker.hpp"
+#include <memory>
+#include <queue>
+#include <thread>
+#include <vector>
+
+class Dispatcher {
+public:
+  Dispatcher() = default;
+  ~Dispatcher() = default;
+
+  Dispatcher(const Dispatcher &) = delete;
+  Dispatcher(Dispatcher &&) = delete;
+  Dispatcher &operator=(const Dispatcher &) = delete;
+  Dispatcher &operator=(Dispatcher &&) = delete;
+
+public:
+  static bool init(const int workers);
+  static void stop();
+  static void addRequest(std::unique_ptr<AbstractRequest> request);
+  static bool addWorker(Worker *worker);
+
+private:
+  static std::queue<std::unique_ptr<AbstractRequest>> requests;
+  static std::queue<Worker *> workers;
+  static std::mutex requests_mtx;
+  static std::mutex workers_mtx;
+  static std::vector<std::unique_ptr<Worker>> all_workers;
+  static std::vector<std::thread> all_threads;
```

Comment:
아마 싱글턴 패턴때문에 static으로 다 선언한거 같은데, 이게 초기화하기도 귀찮고 프로그램 종료까지 데이터 영역에 남아있어서 메모리 관리도 쉽지 않음. 

static을 정말 필요할때만 써줘야 가독성이 올라가긴 함. 
이 함수는 static으로 선언되었기 때문에, class를 그저 namespace처럼 사용하는구나, 또는
이 변수는 static으로 선언되었기 때문에, 통계 같은 걸로 사용되겠구나. 이정도 .


```suggestion
static Dispatcher& getDispatcher () {
     static Dispatcher inst;
     return inst;
     }

public:
  bool init(const int workers);
  void stop();
  void addRequest(std::unique_ptr<AbstractRequest> request);
  bool addWorker(Worker *worker);

private:
  std::queue<std::unique_ptr<AbstractRequest>> requests;
  std::queue<Worker *> workers;
  std::mutex requests_mtx;
  std::mutex workers_mtx;
  std::vector<std::unique_ptr<Worker>> all_workers;
  std::vector<std::thread> all_threads;
```